iroh-net/src/endpoint.rs

```diff
@@ -0,0 +1,245 @@
+    }
+
+    pub async fn bind_with_server_config(
+        keypair: Keypair,
+        bind_addr: SocketAddr,
+        server_config: quinn::ServerConfig,
+        derp_map: Option<DerpMap>,
+        keylog: bool,
+    ) -> anyhow::Result<Self> {
+        let (endpoints_update_s, endpoints_update_r) = flume::bounded(1);
+        let conn = hp::magicsock::Conn::new(hp::magicsock::Options {
+            port: bind_addr.port(),
+            private_key: keypair.secret().clone().into(),
+            on_endpoints: Some(Box::new(move |eps| {
+                if !endpoints_update_s.is_disconnected() && !eps.is_empty() {
+                    endpoints_update_s.send(()).ok();
+                }
+            })),
+            ..Default::default()
+        })
+        .await?;
+        trace!("created magicsock");
+
+        let derp_map = derp_map.unwrap_or_default();
+        conn.set_derp_map(Some(derp_map))
+            .await
+            .context("setting derp map")?;
+
+        let endpoint = quinn::Endpoint::new_with_abstract_socket(
+            quinn::EndpointConfig::default(),
+            Some(server_config),
+            conn.clone(),
+            Arc::new(quinn::TokioRuntime),
+        )?;
+        trace!("created quinn endpoint");
+
+        // Wait for a single endpoint update, to make sure
+        // we found some endpoints
+        tokio::time::timeout(ENDPOINT_WAIT, async move {
+            endpoints_update_r.recv_async().await
+        })
+        .await
+        .context("waiting for endpoint")??;
+
+        Ok(Self {
+            keypair: Arc::new(keypair),
+            conn,
+            endpoint,
+            netmap: Arc::new(Mutex::new(netmap::NetworkMap { peers: vec![] })),
+            keylog,
+        })
+    }
+
+    pub fn accept(&self) -> quinn::Accept<'_> {
+        self.endpoint.accept()
+    }
+
+    pub fn peer_id(&self) -> PeerId {
+        self.keypair.public().into()
+    }
+
+    pub async fn connect(
+        &self,
+        peer_id: PeerId,
+        alpn: &[u8],
+        addrs: &[SocketAddr],
+    ) -> anyhow::Result<quinn::Connection> {
+        self.add_known_addrs(peer_id, addrs).await?;
+
+        let node_key: hp::key::node::PublicKey = peer_id.into();
+        let addr = self
+            .conn
+            .get_mapping_addr(&node_key)
+            .await
+            .expect("just inserted");
+
+        let client_config = {
+            let alpn_protocols = vec![alpn.to_vec()];
+            let tls_client_config =
+                tls::make_client_config(&self.keypair, Some(peer_id), alpn_protocols, self.keylog)?;
+            let mut client_config = quinn::ClientConfig::new(Arc::new(tls_client_config));
+            let mut transport_config = quinn::TransportConfig::default();
+            transport_config.keep_alive_interval(Some(Duration::from_secs(1)));
+            client_config.transport_config(Arc::new(transport_config));
+            client_config
+        };
```

Comment:
This block could also be done at `quinn::Endpoint` creation time with `Endpoint::set_default_client_config`.  In this code-layout this doesn't matter and may even be harder to follow, but I'm not sure if wrapping the endpoint is the right approach.  See my top-level comment.

Reply:
Can it really? The crypto config contains the `peer_id` that you're connecting to.

Reply:
oh, i missed that :(